### PR TITLE
[MRG] Pass check_estimator form scikit-learn

### PR DESCRIPTION
**Pull request by NicolasHug**
Most of the checks are passing, ~but still WIP.~

### CODE REVIEW

File: tests/test_gradient_boosting.py [hunk 1]
```diff
@@ -192,6 +192,12 @@ def custom_check_estimator(Estimator):
             # X is both Fortran and C aligned and numba can't compile.
             # Opened numba issue 3569 (not sure if this is by design)
             continue
+        if check is estimator_checks.check_classifiers_train:
+            continue # probas don't exactly sum to 1 (very close though)
```
Comment (ogrisel) [inline tests/test_gradient_boosting.py]: wouldn't it be possible to post process the softmax to renormalize instead?

Comment (NicolasHug) [reply tests/test_gradient_boosting.py]: I guess... but that's some extra computation just for the test to pass.

The sum of probas are really close to 1:

Changing `assert np.sum(y_prob, axis=1), np.ones(n_samples))` to `assert_array_almost_equal(np.sum(y_prob, axis=1), np.ones(n_samples))` 
make the check pass.

Alternatively we could switch to float64...

File: pygbm/gradient_boosting.py [hunk 2]
```diff
@@ -99,6 +100,7 @@ def fit(self, X, y):
         rng = check_random_state(self.random_state)
 
         self._validate_parameters()
+        self.n_features_ = X.shape[1]  # used for validation in predict()
```
Comment (ogrisel) [inline pygbm/gradient_boosting.py]: Instead of introducing a new attribute in the public API I would rather add a `n_features` attribute to the `TreePredictor` class to be passed as a constructor parameter and use that to do the check in the fitted `GradientBoostingMachine` classes.

File: tests/test_gradient_boosting.py [hunk 3]
```diff
@@ -171,3 +172,55 @@ def test_early_stopping_classification(data, scoring, validation_split, tol):
...
+
+    estimator = Estimator
+    name = type(estimator).__name__
+
+    for check in _yield_all_checks(name, estimator):
+        if (check is estimator_checks.check_fit2d_1feature or
+                check is estimator_checks.check_fit2d_1sample):
+            # X is both Fortran and C aligned and numba can't compile.
+            # Opened numba issue 3569 (not sure if this is by design)
+            continue
```
Comment (ogrisel) [inline tests/test_gradient_boosting.py]: In the meantime, we could make the `GradientBoostingMachine` classes raise a `ValueError` with an informative error message.

File: tests/test_gradient_boosting.py (hunk 1 above)
Comment (ogrisel) [reply tests/test_gradient_boosting.py]: Alright. Maybe we could think of relaxing this check in scikit-learn when `predict_proba` has 32-bit dtype. But this is fine for now.

File: tests/test_gradient_boosting.py [hunk 4]
```diff
@@ -171,3 +172,55 @@ def test_early_stopping_classification(data, scoring, validation_split, tol):
...
+        if (hasattr(check, 'func') and
+                check.func is estimator_checks.check_classifiers_train):
+            continue # same, wrapped in a functools.partial object.
+
+        try:
+            check(name, estimator)
+        except SkipTest as exception:
+            # the only SkipTest thrown currently results from not
+            # being able to import pandas.
+            warnings.warn(str(exception), SkipTestWarning)
```
Comment (ogrisel) [inline tests/test_gradient_boosting.py]: Please add pandas to the `tox.ini` configuration so that we actually run the test with pandas. But ok to not make pandas a mandatory dependency.

File: pygbm/gradient_boosting.py (hunk 2 above)
Comment (NicolasHug) [reply pygbm/gradient_boosting.py]: The problem is that `raw_predict` (where we check the number of features) might be called while no `TreePredictor` has been created yet: we always compute the training / validation score of the init model, before the first boosting iteration.

Comment (ogrisel) [reply pygbm/gradient_boosting.py]: Alright.
